relentless/optimize/method.py

```diff
@@ -121,11 +131,144 @@ def abs_tol(self, value):
+    the minimum of the function. (Does not necessarily converge to minimum.)
+
+    A search interval :math:`\mathbf{d}` is specified, and normalized to :math:`\hat{\mathbf{d}}`:
+
+    .. math::
+
+        \mathbf{d} = \mathbf{x}_{end} - \mathbf{x}_{start}
+
+        \hat{\mathbf{d}} = \frac{\mathbf{d}}{\lVert\mathbf{d}\rVert}
+
+    Then, a 'target' value :math:`t` is defined as:
+
+    .. math::
+
+        t = -\hat{\mathbf{d}} \cdot \nabla{f\left(\mathbf{x}\right)}
+
+    The target at the start of the search interval is always negative. If the
+    target is negative (or within the tolerance) at the end of the search
+    interval, then the maximum step size is acceptable and the algorithm steps
+    to the end of the search interval. If the target is positive (outside of the
+    tolerance) at the end of the search interval, then the algorithm iteratively
+    computes a new step size by linear interpolation within the search interval
+    until the target at the new location is minimized to within the tolerance.
+
+    Parameters
+    ----------
+    abs_tol : float
+        The absolute tolerance for the target.
+    max_iter : int
+        The maximum number of line search iterations allowed.
+
+    """
+    def __init__(self, abs_tol, max_iter):
+        self.abs_tol = abs_tol
+        self.max_iter = max_iter
+
+    def find(self, objective, start, end):
+        """Apply the line search algorithm to take the optimal step.
+
+        Note that the objective function is kept at its initial state, and the
+        function evaluted after taking the optimal step is returned separately.
+
+        Parameters
+        ----------
+        objective : :class:`~relentless.optimize.ObjectiveFunction`
+            The objective function for which the line search is applied.
+        start : :class:`~relentless.optimize.ObjectiveFunctionResult`
+            The objective function evaluated at the start of the search interval.
+        end : :class:`~relentless.optimize.ObjectiveFunctionResult`
+            The objective function evaluated at the end of the search interval.
+
+        Returns
+        -------
+        :class:`~relentless.optimize.ObjectiveFunctionResult`
+            The objective function evaluated at the new, 'optimal' location.
+
+        """
+        ovars = {x: x.value for x in objective.design_variables()}
+
+        # compute normalized search direction
+        d = end.design_variables - start.design_variables
+        max_step = d.norm()
+        d /= max_step
+
+        # compute start and end target values
+        targets = np.array([-d.dot(start.gradient), -d.dot(end.gradient)])
+
+        # check if max step size acceptable, else iterate to minimize target
+        if targets[1] > 0 or np.abs(targets[1]) < self.abs_tol:
+            result = end
+        else:
+            steps = np.array([0, max_step])
+            iter_num = 0
+            new_target = self.abs_tol
```

Comment:
I see what you're doing... better set it to something like `np.inf` so the intention is clear.

Maybe also initialize `new_res = targets[1]` here? That way if the loop is never entered, `new_res` is still defined.